lib/build-manager.js

```diff
@@ -236,6 +244,133 @@ class BuildManager {
+    const vers = await this.getVersionRecord(name, version);
+    if (!vers) {
+      return await this.Version.create({
+        name: name,
+        version: version,
+        value: JSON.stringify(payload)
+      });
+    }
+    throw errs.create({
+      status: 403,
+      message: 'version record already exists'
+    });
+  }
+
+  async getVersionRecord(name, version) {
+    return await this.Version.get({ name, version });
+  }
+
+  async addPackageRecord(pkg) {
+    await this.Package.update(pkg);
+  }
+
+  /**
+  *
+  * untars content into untarPath
+  * @function untarPackage
+  * @param {string} contentData - base64 endcoded content
+  * @param {string} untarPath - path to untar
+  * @returns {function} function to untar
+  */
+  untarPackage(contentData, untarPath) {
+    const stream = through();
+    return new Promise((resolve, reject) => {
+      stream
+        .pipe(zlib.Unzip()) // eslint-disable-line new-cap
+        .once('error', reject)
+        .pipe(tar.extract(untarPath))
+        .once('error', reject)
+        .once('finish', resolve);
+      stream.end(Buffer.from(contentData, 'base64'));
+    });
+  }
+
+  /**
+  *
+  * writes a list of base64 encoded assets to path
+  * @function writeToPath
+  * @param {Object} builtAssets - { [fileName]: { data: [base64 endcoded content] } }
+  * @param {String} unpackPath - path to unpack
+  * @returns {function} function to unpack
+  */
+  async writeToPath(builtAssets, unpackPath) {
+    await fsPromises.mkdir(unpackPath, { recursive: true });
+    return Promise.all(Object.entries(builtAssets).map(([fileName, fileContent]) => {
+      const fileData = fileContent.data;
+      const writePath = path.join(unpackPath, fileName);
+      const buffer = Buffer.from(fileData, 'base64');
+      return fsPromises.writeFile(writePath, buffer);
+    }));
+  }
+  /**
+  *
+  * Walks directory and generate a list of file infos to pass
+  * to bffs.publish for all non-compressed files
+  * @function walkAndGenerateFileInfos
+  * @param {string} directory - directory
+  * @param {array} files - array of files to add to
+  * @returns {function} function to walk and generate
+  */
+  async walkAndGenerateFileInfos(directory) {
+    const basePath = path.resolve(directory);
+    const removeSubDirs = (stats) => {
+      return !stats.isDirectory();
+    };
+    // Get all relevant paths (including sub-directory files), compressed and non compressed
+    const allPaths  = await readdir.async(directory, { basePath, deep: true, filter: removeSubDirs });
+    // Create a lookup map for gz files
+    const gzFileMap = new Map(
+      allPaths.filter(fp => path.extname(fp) === '.gz').map(fp => {
+        return [fp.slice(0, fp.length - 3), fp];
+      })
+    );
+
+    const nonCompressedFilePaths = allPaths.filter(fp => path.extname(fp) !== '.gz');
+
+    const fileInfos = await Promise.all(nonCompressedFilePaths.map(async fp => {
+      const content = await fsPromises.readFile(fp);
```

Comment:
id just make sure that this is working they way we expect, the ID should be verified in the tests.

Reply:
could you clarify what you mean by ID? The fingerprint? is unit testing the walkAndGenerateFileInfos for a test directory with some sample files inside, and asserting the expected fileInfos enough?

Reply:
@jcrugzz not sure how to do the alternative to return a readFile to get the content that seems still async to me?